File: cpp/118.Pascal's Triangle.cpp
/**
 * Source: LeetCode  118. Pascal's Triangle
 * Author: ceezyyy
 * Date: 05-07-2019
 * Reference: @wcxdell https://blog.csdn.net/a2331046/article/details/51538297
 */


/*
Easy

Share
Given a non-negative integer numRows, generate the first numRows of Pascal’s triangle.
In Pascal’s triangle, each number is the sum of the two numbers directly above it.

Example:
Input: 5
Output:
[
[1],
[1,1],
[1,2,1],
[1,3,3,1],
[1,4,6,4,1]
]

*/


#include<vector>
using namespace std;

class Solution {
public:
	vector<vector<int>> generate(int numRows) {
		vector<vector<int>> res; // 2D array
		for (int i = 0; i < numRows; i++) {
			vector<int> sub(i + 1); // the length of each subarray is different
			sub[0] = sub[i] = 1; // the first & last '1'
			for (int j = 1; j < i; j++) {
				sub[j] = res[i - 1][j - 1] + res[i - 1][j];
			}
			res.push_back(sub); // join each subarray
		}
		return res;
	}
};


// Study Notes:
// 2D array & subarray 


/**
 * Submission Detail:
 * Runtime: 4 ms, faster than 100.00% of C++ online submissions for Pascal's Triangle.
 * Memory Usage: 9 MB, less than 12.17% of C++ online submissions for Pascal's Triangle.
 */

File: others/Python3/118.Pascal's Triangle.py
"""
@Title: 118. Pascal's Triangle
@Tag: dp
@Date: Dec-31 2019
@Author: ceezyyy
@Difficulty: Easy
"""


class Solution:
    def generate(self, numRows: int) -> List[List[int]]:
        # memorization (set default as one)
        dp = [[1 for j in range(i+1)] for i in range(numRows)]
        for row in range(2, numRows):  # start from the row[2]
            for col in range(1, row):
                dp[row][col] = dp[row-1][col-1]+dp[row-1][col]
        return dp


"""
Runtime: 28 ms, faster than 81.27% of Python3 online submissions for Pascal's Triangle.
Memory Usage: 12.7 MB, less than 100.00% of Python3 online submissions for Pascal's Triangle.
"""


"""
Time Complexity: O(m*n)
Space Complexity: O(m*n)
"""
